### PR TITLE
add hints from solution (+ clippy fixes)

### PR DESCRIPTION
**Pull request by FrederickFrance**

### CODE REVIEW

File: src/builder.rs [hunk 1]
```diff
@@ -594,6 +594,11 @@ impl CpModelBuilder {
         Constraint(index)
     }
 
+    /// Delete all constraints.
+    pub fn del_all_cst(&mut self) {
```
Comment (TeXitoi) [inline src/builder.rs:598]: I prefer to not add this method: the cpp interface doesn't propose it, and it's quite dangerous in a real modeling situation, as you might think some variable still has their values, and is not really useful in practice as you can just create a new model.

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "cp_sat"
-version = "0.3.0"
+version = "0.4.0"
```
Comment (TeXitoi) [inline Cargo.toml:3]: no breaking changes => 0.3.1

File: src/builder.rs [hunk 3]
```diff
@@ -624,6 +629,48 @@ impl CpModelBuilder {
         }
     }
 
+    /// Add all solution hints from a solution.
```
Comment (TeXitoi) [inline src/builder.rs:632]: providing a solution with a vector is very error prone, as you cannot really know the order of the variables.

File: src/builder.rs [hunk 4]
```diff
@@ -624,6 +629,48 @@ impl CpModelBuilder {
...
+    /// assert_eq!(response.status(), CpSolverStatus::Optimal);
+    ///  
+    /// ```
+    pub fn add_hints(&mut self, solution: Vec<i64>) {
+        let hints = self
+            .proto
+            .solution_hint
+            .get_or_insert_with(Default::default);
+
+        for (i, _) in solution.iter().enumerate().take(self.proto.variables.len()) {
```
Comment (TeXitoi) [inline src/builder.rs:663]: And this function will not work, as if you have done `add_hints`  before, they will not be erased.

File: src/builder.rs [hunk 5]
```diff
@@ -624,6 +629,48 @@ impl CpModelBuilder {
...
+            hints.vars.push(i as i32);
+            hints.values.push(solution[i]);
+        }
+    }
+
+    /// Delete all solution hints.
```
Comment (TeXitoi) [inline src/builder.rs:669]: Need an example

**Review by TeXitoi — CHANGES_REQUESTED** [review]

File: src/builder.rs (hunk 1 above)
Comment (FrederickFrance) [reply src/builder.rs]: Ok

File: Cargo.toml (hunk 2 above)
Comment (FrederickFrance) [reply Cargo.toml]: Ok

File: src/builder.rs (hunk 3 above)
Comment (FrederickFrance) [reply src/builder.rs]: Ok

File: src/builder.rs (hunk 4 above)
Comment (FrederickFrance) [reply src/builder.rs]: Ok

File: src/builder.rs (hunk 5 above)
Comment (FrederickFrance) [reply src/builder.rs]: done

**Review by TeXitoi — APPROVED** [review]

**Pull request merged**